File: scripts/overture-footprints.py
#!/usr/bin/env python3
"""Extract the building footprints OpenStreetMap has not mapped yet.

WHY THIS IS A BUILD STEP AND NOT A FETCH
========================================
Overture Maps publishes the planet's buildings as GeoParquet on public S3 — free,
no key, no account. Getting the 714 buildings of one Shanghai bbox took DuckDB
140 seconds of columnar scanning. No browser is doing that while somebody waits,
so the districts we care about are extracted once, here, and shipped as a small
JSON file the viewer loads beside its Overpass query.

WHAT IS FILTERED HERE, AND WHAT IS NOT
======================================
Overture publishes each building's provenance, and this script keeps only the
rows it does NOT attribute to OpenStreetMap — those we already have, and better,
because they are hand-mapped and carry tags.

That filter is necessary and *not sufficient*. Measured over the Lujiazui core
bbox: of 308 ML-sourced rows, **47 (15%) land on an OSM building anyway**,
because Overture kept both where its own matcher failed to pair them. So a
second, geometric pass is required — and it deliberately lives at RUNTIME in
`src/lib/geo/overture-footprints.ts`, not here. OSM grows; a de-duplication
baked in at build time goes stale against exactly the data it complements.

HEIGHTS ARE NOT TAKEN
=====================
Overture's China heights are OSM-derived, so its ML footprints carry almost
none: 7 of 287 in the core bbox. Reading a height here would import a guess we
cannot inspect. These buildings arrive with no height and are given the
district's own measured median by the local height prior, flagged estimated.

Usage:
    pip install duckdb
    python scripts/overture-footprints.py

Writes public/geo/overture/<slug>.json for each district below.
"""

from __future__ import annotations

import io
import json
import math
import os
import sys
import time

# Overture release to pin. Bumping this is a deliberate act: a new release can
# add, move or withdraw footprints, and the extract should change when somebody
# decided it should, not because a build ran on a Tuesday.
RELEASE = "2026-08-19.0"

S3 = f"s3://overturemaps-us-west-2/release/{RELEASE}/theme=buildings/type=building/*.parquet"

ATTRIBUTION = "© Overture Maps Foundation"

# Half-size of each extract, metres. Matches BUILDINGS_HALF_SIZE_M in
# geo-system so the extract covers exactly what the Overpass query does —
# a larger one ships bytes nobody draws, a smaller one leaves a visible edge
# where the extra buildings stop and OSM's carry on.
HALF_SIZE_M = 700

DISTRICTS = [
    # slug,            lat,        lon,        label
    ("lujiazui",       31.2397,   121.4998,   "Lujiazui, Pudong"),
    ("poblenou",       41.4064,     2.1900,   "Poblenou, Barcelona"),
    ("barceloneta",    41.3687,     2.1900,   "Barceloneta / Port Vell"),
]

OUT_DIR = os.path.join("public", "geo", "overture")


def bbox_around(lat: float, lon: float, half_m: float):
    d_lat = half_m / 111_132.0
    d_lon = half_m / (111_320.0 * math.cos(math.radians(lat)))
    return (lon - d_lon, lat - d_lat, lon + d_lon, lat + d_lat)


def extract(con, slug: str, lat: float, lon: float, label: str) -> dict:
    west, south, east, north = bbox_around(lat, lon, HALF_SIZE_M)
    t0 = time.time()
    rows = con.execute(
        f"""
        SELECT id, ST_AsGeoJSON(geometry) AS gj, sources[1].dataset AS src
        FROM read_parquet('{S3}')
        WHERE bbox.xmin > {west} AND bbox.xmax < {east}
          AND bbox.ymin > {south} AND bbox.ymax < {north}
        """
    ).fetchall()
    took = time.time() - t0

    kept, from_osm, unusable = [], 0, 0
    for bid, gj, src in rows:
        # The provenance filter. OSM-sourced rows are ones we already fetch, and
        # hand-mapped with tags, so ours are strictly better.
        if src == "OpenStreetMap":
            from_osm += 1
            continue
        try:
            geom = json.loads(gj)
        except Exception:
            unusable += 1
            continue
        # Outer ring only. A courtyard is a hole we cannot express in the
        # viewer's footprint model, and filling it is better than dropping the
        # building — but taking an INNER ring as a building would draw the void.
        coords = None
        if geom.get("type") == "Polygon":
            coords = geom["coordinates"][0]
        elif geom.get("type") == "MultiPolygon" and geom["coordinates"]:
            coords = max(geom["coordinates"], key=lambda p: len(p[0]))[0]
        if not coords or len(coords) < 4:
            unusable += 1
            continue
        # GeoJSON closes its rings; the viewer's model does not.
        ring = [{"lat": round(c[1], 7), "lon": round(c[0], 7)} for c in coords[:-1]]
        kept.append({"id": str(bid), "ring": ring})

    print(
        f"  {slug:14s} {len(rows):5d} rows in {took:5.0f}s"
        f"  ->  {from_osm:4d} already OSM, {unusable:3d} unusable, {len(kept):4d} shipped"
    )
    return {
        "release": RELEASE,
        "attribution": ATTRIBUTION,
        "label": label,
        "bbox": [round(west, 6), round(south, 6), round(east, 6), round(north, 6)],
        "footprints": kept,
    }


def main() -> int:
    try:
        import duckdb  # noqa: PLC0415
    except ImportError:
        print("duckdb is required:  pip install duckdb", file=sys.stderr)
        return 1

    con = duckdb.connect()
    con.execute("INSTALL httpfs; LOAD httpfs; SET s3_region='us-west-2';")
    con.execute("INSTALL spatial; LOAD spatial;")

    os.makedirs(OUT_DIR, exist_ok=True)
    print(f"Overture release {RELEASE}, {HALF_SIZE_M} m half-size\n")
    index = []
    for slug, lat, lon, label in DISTRICTS:
        data = extract(con, slug, lat, lon, label)
        path = os.path.join(OUT_DIR, f"{slug}.json")
        with io.open(path, "w", encoding="utf-8") as fh:
            json.dump(data, fh, separators=(",", ":"))
        print(f"  {'':14s} wrote {path}  ({os.path.getsize(path) / 1024:.0f} KB)")
        index.append(
            {"slug": slug, "label": label, "bbox": data["bbox"],
             "count": len(data["footprints"])}
        )

    # An index, so the viewer asks one small question — "is there an extract for
    # where this model stands?" — before deciding to download a district.
    # Without it every site would have to guess a slug, and a 404 on every load
    # is a worse answer than a 1 KB list.
    index_path = os.path.join(OUT_DIR, "index.json")
    with io.open(index_path, "w", encoding="utf-8") as fh:
        json.dump(
            {"release": RELEASE, "attribution": ATTRIBUTION, "districts": index},
            fh, separators=(",", ":"),
        )
    print("")
    print(f"  wrote {index_path}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/lib/geo/overture-footprints.ts
// ─── overture-footprints ──────────────────────────────────────────────────────
// THE BUILDINGS OSM HAS NOT MAPPED YET.
//
// OpenStreetMap maps 937 building outlines in the Lujiazui benchmark patch.
// Overture, over the same bbox, has 1 494 — the same 406 in the core area plus
// several hundred detected by machine learning. Those extra footprints are real
// buildings that are simply not in OSM, and a district drawn without them has
// visible gaps where a city has blocks.
//
// ── Why a shipped extract and not a live query ────────────────────────────────
//
// Overture is GeoParquet on public S3. Getting the 714 buildings of one bbox
// took DuckDB 140 seconds of columnar scanning; no browser is going to do that
// while somebody waits. So the districts we care about are extracted once at
// build time by `scripts/overture-footprints.py` and shipped as a small file.
//
// ── Why the de-duplication lives HERE and not in the extract ──────────────────
//
// Overture publishes each building's provenance, so the extract can drop
// everything it attributes to OpenStreetMap. That is necessary and NOT
// sufficient, which measuring is what showed: of the 308 ML-sourced rows in the
// core bbox, **47 — 15% — land on an OSM building anyway**, because Overture
// kept both where its own matcher failed to pair them. Shipping those would
// draw 47 buildings twice, one slightly beside the other, which is worse than
// the gaps it set out to fill.
//
// So a second, geometric pass is required, and it belongs at RUNTIME: OSM grows,
// and a de-duplication baked in at build time goes stale against exactly the
// data it exists to complement. A footprint suppressed today because OSM has it
// must stay suppressed tomorrow when OSM moves it slightly; one admitted today
// must disappear the day somebody maps it.
//
// ── What these buildings do NOT bring ─────────────────────────────────────────
//
// Heights. Overture's China heights are OSM-derived, so the ML footprints carry
// none: 7 of the 287 in the core bbox, against 39 for the whole set. Every one
// of these arrives `estimated`, which is why this was worth doing only after
// the local height prior landed — before it they would all have been 8 m boxes.
//
// PURE: footprints and OSM rings in, the subset worth drawing out. No fetch, no
// THREE, no tags.

/** One footprint as shipped: a closed ring, and nothing else. */
export interface OvertureFootprint {
  id: string
  /** Outer ring, degrees. Not closed — the last point does not repeat. */
  ring: Array<{ lat: number; lon: number }>
}

/** The file `scripts/overture-footprints.py` writes. */
export interface OvertureExtract {
  /** Overture release the extract came from, e.g. `2026-08-19.0`. */
  release: string
  /** Attribution the viewer must display while these are on screen. */
  attribution: string
  /** [west, south, east, north] the extract covers. */
  bbox: [number, number, number, number]
  footprints: OvertureFootprint[]
}

/**
 * How near an Overture centroid must be to an OSM building to count as the same
 * one, metres.
 *
 * ML footprints are traced from imagery and OSM ones by hand from the same
 * imagery, so the two disagree by roughly the width of a wall rather than by
 * the width of a building. Generous enough to catch that; tight enough that two
 * genuinely adjacent town houses do not collapse into one.
 */
export const DUPLICATE_RADIUS_M = 8

/** Anything this small is imagery noise, not a building. */
export const MIN_AREA_M2 = 25

export interface RingLike {
  ring: ReadonlyArray<{ lat: number; lon: number }>
}

interface Placed {
  lon: number
  lat: number
  west: number
  east: number
  south: number
  north: number
}

function place(ring: ReadonlyArray<{ lat: number; lon: number }>): Placed | null {
  if (ring.length < 3) return null
  let lon = 0
  let lat = 0
  let west = Infinity
  let east = -Infinity
  let south = Infinity
  let north = -Infinity
  for (const p of ring) {
    if (!Number.isFinite(p.lat) || !Number.isFinite(p.lon)) return null
    lon += p.lon
    lat += p.lat
    if (p.lon < west) west = p.lon
    if (p.lon > east) east = p.lon
    if (p.lat < south) south = p.lat
    if (p.lat > north) north = p.lat
  }
  return { lon: lon / ring.length, lat: lat / ring.length, west, east, south, north }
}

/** Shoelace area in m², local equirectangular. Enough to reject slivers. */
export function ringAreaM2(ring: ReadonlyArray<{ lat: number; lon: number }>): number {
  if (ring.length < 3) return 0
  const kx = 111_320 * Math.cos((ring[0].lat * Math.PI) / 180)
  const ky = 111_132
  let s = 0
  for (let i = 0; i < ring.length; i++) {
    const a = ring[i]
    const b = ring[(i + 1) % ring.length]
    s += a.lon * kx * (b.lat * ky) - b.lon * kx * (a.lat * ky)
  }
  return Math.abs(s) / 2
}

/**
 * The Overture footprints worth adding, given what OSM already has here.
 *
 * A footprint is dropped when its centroid falls INSIDE an OSM building's
 * bounding box, or within `DUPLICATE_RADIUS_M` of an OSM centroid. Two tests
 * rather than one because they catch different failures: a large OSM building
 * whose centroid is far from a small overlapping ML footprint is caught by the
 * box, and two small footprints offset by a metre are caught by the radius.
 *
 * Deliberately conservative in one direction. A false DROP costs one building
 * the city already had; a false KEEP draws a building twice, slightly offset,
 * which reads as a rendering fault rather than as missing data.
 */
export function newFootprints(
  extract: ReadonlyArray<OvertureFootprint>,
  osm: ReadonlyArray<RingLike>,
  radiusM = DUPLICATE_RADIUS_M,
): OvertureFootprint[] {
  const placed = osm
    .map((b) => place(b.ring))
    .filter((p): p is Placed => p !== null)
  if (extract.length === 0) return []

  const out: OvertureFootprint[] = []
  for (const f of extract) {
    const p = place(f.ring)
    if (!p) continue
    if (ringAreaM2(f.ring) < MIN_AREA_M2) continue

    // Metres per degree at this latitude, computed per footprint rather than
    // once for the patch: a district can be tall enough for the cosine to move.
    const kx = 111_320 * Math.cos((p.lat * Math.PI) / 180)
    const ky = 111_132

    let duplicate = false
    for (const o of placed) {
      if (p.lon >= o.west && p.lon <= o.east && p.lat >= o.south && p.lat <= o.north) {
        duplicate = true
        break
      }
      const dx = (p.lon - o.lon) * kx
      const dy = (p.lat - o.lat) * ky
      if (dx * dx + dy * dy <= radiusM * radiusM) {
        duplicate = true
        break
      }
    }
    if (!duplicate) out.push(f)
  }
  return out
}

/** Does this extract cover the point a model was placed at? */
export function extractCovers(
  extract: Pick<OvertureExtract, 'bbox'>, lat: number, lon: number,
): boolean {
  const [w, s, e, n] = extract.bbox
  return lon >= w && lon <= e && lat >= s && lat <= n
}
